### PR TITLE
Refresh control test

### LINKED ISSUE #19820: Wallet: Pull to refresh - Home Page
Add an implementation for pull to refresh on the home page.

Please align with one other developer when working on this as the solution should be somewhat reusable to work well with other parts of the application, e.g account page/ various other lists in the application.

Designs:
https://www.figma.com/file/xLs1KYmF4e6WwRTZVJKeUK/Descoped-Wallet?type=design&node-id=7503-250396&mode=design&t=46E6NuWhgSJA2dqj-4

See a demo branch from @ulisesmac with some initial work on this:
`refresh-control-test` - https://github.com/status-im/status-mobile/tree/refresh-control-test

### PR DESCRIPTION
**Pull request by mmilad75**
fixes #19820

UPDATED:

https://github.com/status-im/status-mobile/assets/55688834/da0b44d3-7b26-4eac-9c44-9072b3daa51c

Android:

** note that the list of tokens have been tripled manually for test purposes. 

https://github.com/status-im/status-mobile/assets/55688834/5fe8b547-c6c7-4676-9b2f-803db56e8d43

### CODE REVIEW

File: src/status_im/contexts/wallet/home/view.cljs [hunk 1]
```diff
@@ -51,30 +54,42 @@
...
-       :size           32
-       :default-active selected-tab
-       :data           tabs-data
-       :on-change      #(set-selected-tab %)}]
-     [tabs/view {:selected-tab selected-tab}]]))
+     [rn/scroll-view
+      {:refresh-control (reagent/as-element
+                         [rn/refresh-control
+                          {:refreshing (and tokens-loading? init-loaded?)
+                           :colors     colors/neutral-40
```
Comment (J-Son89) [inline src/status_im/contexts/wallet/home/view.cljs]: should these colors be themed?

File: src/status_im/contexts/wallet/home/view.cljs [hunk 2]
```diff
@@ -51,30 +54,42 @@
...
-       :shows-horizontal-scroll-indicator false}]
-     [quo/tabs
-      {:style          style/tabs
-       :size           32
-       :default-active selected-tab
-       :data           tabs-data
-       :on-change      #(set-selected-tab %)}]
-     [tabs/view {:selected-tab selected-tab}]]))
+     [rn/scroll-view
+      {:refresh-control (reagent/as-element
```
Comment (J-Son89) [inline src/status_im/contexts/wallet/home/view.cljs]: I wonder can we abstract this as-element away?

File: src/status_im/contexts/wallet/home/view.cljs (hunk 1 above)
Comment (mmilad75) [reply src/status_im/contexts/wallet/home/view.cljs]: I couldn't find the design for dark theme, so I used this for both, and it looked visible and good

File: src/status_im/contexts/wallet/home/view.cljs (hunk 2 above)
Comment (mmilad75) [reply src/status_im/contexts/wallet/home/view.cljs]: I removed it but I got errors. @ulisesmac  added this and I kept it.

Comment (J-Son89) [reply src/status_im/contexts/wallet/home/view.cljs]: Im not saying to remove it, I mean to abstract it away so we don't need to add this method each time we use the pull to refresh,  there is many other uses of pull to refresh in the designs so it's nice to take steps like this

Comment (mmilad75) [reply src/status_im/contexts/wallet/home/view.cljs]: Oh, right. We can create a new refresh-control component and add this there

Comment (J-Son89) [reply src/status_im/contexts/wallet/home/view.cljs]: nice yeah that sounds good to me!

Comment (mmilad75) [reply src/status_im/contexts/wallet/home/view.cljs]: Done

File: src/quo/components/scroll_view/view.cljs [hunk 3]
```diff
@@ -0,0 +1,11 @@
+(ns quo.components.scroll-view.view
```
Comment (briansztamfater) [inline src/quo/components/scroll_view/view.cljs]: nit: maybe `refreshable-scroll-view`?
or add some description of what is different from regular ScrollView

**Comment by briansztamfater** [comment]
I think it is a UX redundancy to show the activity loading indicator and the whole screen loading skeleton while refreshing. IMO the loading skeleton should only be shown the first time because there is no data to show at that time, for refreshing we could just should the activity indicator from refresh control component.

File: src/quo/components/scroll_view/view.cljs (hunk 3 above)
Comment (mmilad75) [reply src/quo/components/scroll_view/view.cljs]: Nice 👍

**Comment by mmilad75** [comment]
> I think it is a UX redundancy to show the activity loading indicator and the whole screen loading skeleton while refreshing. IMO the loading skeleton should only be shown the first time because there is no data to show at that time, for refreshing we could just should the activity indicator from refresh control component.

When trying to re-fetch the accounts using `:wallet/get-accounts`, the accounts balance and other props will be reset and will be 0 in the ui. If we want to skip resetting them, it will require more job. While, @J-Son89 suggested me to do the ticket with minimal changes and efforts. I think it's better to show both the loading indicator and the loading status of the accounts for now.

File: src/quo/components/scroll_view/view.cljs (hunk 3 above)
Comment (mmilad75) [reply src/quo/components/scroll_view/view.cljs]: Done

**Comment by briansztamfater** [comment]
> > I think it is a UX redundancy to show the activity loading indicator and the whole screen loading skeleton while refreshing. IMO the loading skeleton should only be shown the first time because there is no data to show at that time, for refreshing we could just should the activity indicator from refresh control component.
> 
> When trying to re-fetch the accounts using `:wallet/get-accounts`, the accounts balance and other props will be reset and will be 0 in the ui. If we want to skip resetting them, it will require more job. While, @J-Son89 suggested me to do the ticket with minimal changes and efforts. I think it's better to show both the loading indicator and the loading status of the accounts for now.

Sounds good, maybe we should confirm with the design team if they are okay with this behavior for our release or if we need a follow up issue for prioritize and implement in the future.

**Comment by mmilad75** [comment]
> > > I think it is a UX redundancy to show the activity loading indicator and the whole screen loading skeleton while refreshing. IMO the loading skeleton should only be shown the first time because there is no data to show at that time, for refreshing we could just should the activity indicator from refresh control component.
> > 
> > 
> > When trying to re-fetch the accounts using `:wallet/get-accounts`, the accounts balance and other props will be reset and will be 0 in the ui. If we want to skip resetting them, it will require more job. While, @J-Son89 suggested me to do the ticket with minimal changes and efforts. I think it's better to show both the loading indicator and the loading status of the accounts for now.
> 
> Sounds good, maybe we should confirm with the design team if they are okay with this behavior for our release or if we need a follow up issue for prioritize and implement in the future.

I checked with the design team and they approved it (https://discord.com/channels/624634427930312714/928625369542713396/1235661790109503589). But they asked for replacing the default spinner with the one in the designs. However, @J-Son89 suggested me to skip it since it will take time

**Review by OmarBasem — CHANGES_REQUESTED** [review]
Looks nice, but VirtualizedLists (FlatLists) should never be nested inside plain ScrollViews. 

The component needs to be a FlatList where the assets items are the `data` and the rest of the components above it are the FlatList header

FYI, when nesting FlatList inside a ScrollView, the FlatList optimizations regarding rendering children will not be applied and will be treated as a ScrollView. (also you should be getting a warning from react-native in metro about this)

**Comment by mmilad75** [comment]
> Looks nice, but VirtualizedLists (FlatLists) should never be nested inside plain ScrollViews.
> 
> The component needs to be a FlatList where the assets items are the `data` and the rest of the components above it are the FlatList header
> 
> FYI, when nesting FlatList inside a ScrollView, the FlatList optimizations regarding rendering children will not be applied and will be treated as a ScrollView

The data in the scroll view has different render options
In these cases, it should be scroll view
And besides that, @ulisesmac had started this component, and he used scroll view
I don't think it's a good idea to go with FlatList

**Review by briansztamfater — APPROVED** [review]
Looks good for me, FlatList is horizontal to display account cards so I think it is fine to use ScrollView

**Comment by OmarBasem** [comment]
> Looks good for me, FlatList is horizontal to display account cards so I think it is fine to use ScrollView

The problem is not the account cards. We have the ScrollView is vertical and the assets list is vertical.

**Comment by OmarBasem** [comment]
> > Looks nice, but VirtualizedLists (FlatLists) should never be nested inside plain ScrollViews.
> > The component needs to be a FlatList where the assets items are the `data` and the rest of the components above it are the FlatList header
> > FYI, when nesting FlatList inside a ScrollView, the FlatList optimizations regarding rendering children will not be applied and will be treated as a ScrollView
> 
> The data in the scroll view has different render options In these cases, it should be scroll view And besides that, @ulisesmac had started this component, and he used scroll view I don't think it's a good idea to go with FlatList

@mmilad75 as I discussed with you earlier using a FlatList inside a ScrollView will render the FlatList optimizations useless, consume more RAM, and have a negative effect on the scrolling performance with large lists: 

https://stackoverflow.com/a/55256341.

https://reactnative.dev/docs/performance#listview-initial-rendering-is-too-slow-or-scroll-performance-is-bad-for-large-lists.

We plan to make the wallet home our landing page, so let's not rush into a sub-optimal solution. We can discuss this further with other devs.

cc @J-Son89 @ulisesmac

File: src/quo/components/refreshable_scroll_view/view.cljs [hunk 4]
```diff
@@ -0,0 +1,11 @@
+(ns quo.components.refreshable-scroll-view.view
```
Comment (J-Son89) [inline src/quo/components/refreshable_scroll_view/view.cljs]: Quo components should really be only things related to the design system file.

Imo this seems better to have in status-im.common... ns

**Comment by mmilad75** [comment]
> > > Looks nice, but VirtualizedLists (FlatLists) should never be nested inside plain ScrollViews.
> > > The component needs to be a FlatList where the assets items are the `data` and the rest of the components above it are the FlatList header
> > > FYI, when nesting FlatList inside a ScrollView, the FlatList optimizations regarding rendering children will not be applied and will be treated as a ScrollView
> > 
> > 
> > The data in the scroll view has different render options In these cases, it should be scroll view And besides that, @ulisesmac had started this component, and he used scroll view I don't think it's a good idea to go with FlatList
> 
> @mmilad75 as I discussed with you earlier using a FlatList inside a ScrollView will render the FlatList optimizations useless, consume more RAM, and have a negative effect on the scrolling performance with large lists:
> 
> https://stackoverflow.com/a/55256341.
> 
> https://reactnative.dev/docs/performance#listview-initial-rendering-is-too-slow-or-scroll-performance-is-bad-for-large-lists.
> 
> We plan to make the wallet home our landing page, so let's not rush into a sub-optimal solution. We can discuss this further with other devs.
> 
> cc @J-Son89 @ulisesmac

I have converted it (not pushed yet), and the result will be like this. Is this acceptable?

https://github.com/status-im/status-mobile/assets/55688834/56acd7b1-0916-4d46-ad55-b5b98a4a9b64

**Comment by OmarBasem** [comment]
> I have converted it (not pushed yet), and the result will be like this. Is this acceptable?

Looks good @mmilad75!

I think we can move the bar at the top that has the total balance out of the list header and it would be good. Wdyt @J-Son89 ?

Also, I think it was said that we will remove the graph @J-Son89 ?

**Comment by mmilad75** [comment]
> > I have converted it (not pushed yet), and the result will be like this. Is this acceptable?
> 
> Looks good @mmilad75!
> 
> I think we can move the bar at the top that has the total balance out of the list header and it would be good. Wdyt @J-Son89 ?
> 
> Also, I think it was said that we will remove the graph @J-Son89 ?

If we move the top bar to the top, the loading spinner will be under it which is against the designs

**Comment by J-Son89** [comment]
> > I have converted it (not pushed yet), and the result will be like this. Is this acceptable?
> 
> Looks good @mmilad75!
> 
> I think we can move the bar at the top that has the total balance out of the list header and it would be good. Wdyt @J-Son89 ?
> 
> Also, I think it was said that we will remove the graph @J-Son89 ?

where it's position is correct with the designs, but I have hid the graph for now. We should see it in latest develop since I just merged the pr to hide it.

**Comment by mmilad75** [comment]
> > > I have converted it (not pushed yet), and the result will be like this. Is this acceptable?
> > 
> > 
> > Looks good @mmilad75!
> > I think we can move the bar at the top that has the total balance out of the list header and it would be good. Wdyt @J-Son89 ?
> > Also, I think it was said that we will remove the graph @J-Son89 ?
> 
> where it's position is correct with the designs, but I have hid the graph for now. We should see it in latest develop since I just merged the pr to hide it.


https://github.com/status-im/status-mobile/assets/55688834/22244479-3577-4bef-b0d6-94bce5fe7d5f

Is this ok? @J-Son89 @OmarBasem

**Review by OmarBasem — APPROVED** [review]
Looks good @mmilad75! Thanks for your work 🙌

**Comment by qoqobolo** [comment]
Hi @mmilad75 , nice job, thank you!
Take a look at the issue below please:

### ISSUE 1: Assets order changes every time when you refresh the page


https://github.com/status-im/status-mobile/assets/67952253/1995f178-ea26-4805-b8ab-f4ad2ee31865

**Comment by mmilad75** [comment]
> Hi @mmilad75 , nice job, thank you! Take a look at the issue below please:
> 
> ### ISSUE 1: Assets order changes every time when you refresh the page
>  IMG_4690.MP4

The order is not handled on the front end side. I don't think there is anything we can do about it. Backend side should fix it. @qoqobolo 

cc @J-Son89 Please correct me if I'm wrong

**Comment by OmarBasem** [comment]
> > Hi @mmilad75 , nice job, thank you! Take a look at the issue below please:
> > ### ISSUE 1: Assets order changes every time when you refresh the page
> > IMG_4690.MP4
> 
> The order is not handled on the front end side. I don't think there is anything we can do about it. Backend side should fix it. @qoqobolo
> 
> cc @J-Son89 Please correct me if I'm wrong

We can handle the ordering on the front end if we want to. But is this change of ordering after refresh happening on this branch only?

**Comment by qoqobolo** [comment]
> But is this change of ordering after refresh happening on this branch only?

Ah you're right! It's also happening on develop when I reopen the app, I just didn't pay attention to it before tbh.
Sorry for the false alarm then @OmarBasem @mmilad75, please ignore that issue!
I'll continue testing tomorrow.

**Comment by qoqobolo** [comment]
> We can handle the ordering on the front end if we want to.

So by design, it looks like we should control the order: from larger to smaller amount of tokens, I guess

Screenshot 2024-05-08 at 21 19 02


But this is an issue that should be logged separately (if it hasn't been already), and I'm not sure about the priority.
@J-Son89 let me know if we want to include it to the scope

**Comment by qoqobolo** [comment]
> We can handle the ordering on the front end if we want to.

So according to [design](https://www.figma.com/file/xLs1KYmF4e6WwRTZVJKeUK/Descoped-Wallet?type=design&node-id=7503-250396&mode=design&t=wPOke6W250PFqgO4-4), it looks like we should control the order: from larger to smaller amount of tokens, I guess

Screenshot 2024-05-08 at 21 19 02

Which doesn't happen now

Screenshot 2024-05-08 at 21 24 33



But this is an issue that should be logged separately (if it hasn't been already), and I'm not sure about the priority.
@J-Son89 just let me know if we want to include it in the scope

**Comment by J-Son89** [comment]
Thanks @qoqobolo - let's keep this separate!
It should be fine for someone to take provided there is no more high priority bugs.
We are very close to finishing feature development of the wallet for 2.30 release (July).
So beyond a few feature prs we are mostly focusing on bugs and polishing details such as this :)

**Comment by qoqobolo** [comment]
Everything looks good to me, thanks for your work @mmilad75!
PR can be merged.

**Pull request merged**